File: nucleoid/lang/ast/identifier.py
"""
Identifier AST node implementation.
"""

import copy
from collections.abc import Iterator
from typing import Any, Optional

from ..estree.generator import append, root
from .node import Node


class Identifier(Node):
    """
    Identifier node for variable references and member expressions.
    
    Supports complex identifier chains like obj.prop.subprop and provides
    methods for traversal, resolution, and manipulation.
    """

    @staticmethod
    def get_types() -> list[str]:
        """Get supported AST node types for identifiers."""
        return ["Identifier", "MemberExpression", "ThisExpression"]

    @property
    def first(self) -> Optional['Identifier']:
        """
        Get the first (leftmost) identifier in a chain.
        
        For 'obj.prop.sub', returns 'obj'.
        For simple identifiers, returns self.
        """
        node_type = self.node.get("type", "")

        if node_type in ["Identifier", "ThisExpression", "Literal"]:
            return Identifier(self.node)
        elif node_type == "MemberExpression":
            r = root(self.node)
            if r.get("object"):
                return Identifier(r["object"])

        return None

    @first.setter
    def first(self, first: 'Identifier') -> None:
        """Set the first identifier in the chain."""
        current_first = self.first
        if not current_first:
            raise ValueError("Cannot set first on non-MemberExpression")

        # Clear current node and replace with first
        self.node.clear()
        self.node.update(first.node)

    @property
    def object(self) -> Optional['Identifier']:
        """Get the object part of a member expression."""
        node_type = self.node.get("type", "")

        if node_type in ["Identifier", "ThisExpression", "Literal"]:
            return None
        elif node_type == "MemberExpression":
            obj = self.node.get("object")
            if obj:
                return Identifier(obj)

        return None

    @object.setter
    def object(self, obj: 'Identifier') -> None:
        """Set the object part of a member expression."""
        current_object = self.object
        if not current_object:
            raise ValueError("Cannot set object on non-MemberExpression")

        self.node["object"] = obj.node

    @property
    def last(self) -> Optional['Identifier']:
        """
        Get the last (rightmost) identifier in a chain.
        
        For 'obj.prop.sub', returns 'sub'.
        For simple identifiers, returns self.
        """
        node_type = self.node.get("type", "")

        if node_type in ["Identifier", "ThisExpression", "Literal"]:
            return Identifier(self.node)
        elif node_type == "MemberExpression":
            prop = self.node.get("property")
            if prop:
                return Identifier(prop)

        return None

    @last.setter
    def last(self, last: 'Identifier') -> None:
        """Set the last identifier in the chain."""
        current_last = self.last
        if not current_last:
            raise ValueError("Cannot set last on non-MemberExpression")

        if self.node.get("type") == "MemberExpression":
            self.node["property"] = last.node
        else:
            # Replace entire node with last
            self.node.clear()
            self.node.update(last.node)

    def __str__(self) -> str:
        """Convert identifier to string representation."""
        node_type = self.node.get("type", "")

        if node_type == "Identifier":
            return self.node.get("name", "")
        elif node_type == "ThisExpression":
            return "this"
        elif node_type == "MemberExpression":
            obj = self.node.get("object", {})
            prop = self.node.get("property", {})
            obj_str = Identifier(obj).__str__()
            prop_str = Identifier(prop).__str__()
            return f"{obj_str}.{prop_str}"
        elif node_type == "Literal":
            return str(self.node.get("value", ""))

        return ""

    def resolve(self, scope: Any | None = None) -> Any:
        """
        Resolve identifier in the given scope.
        
        Args:
            scope: The scope object to resolve against
            
        Returns:
            The resolved AST node or value
        """
        if not scope:
            return self.node

        first = self.first
        if not first:
            return None

        # Check if this is a callback parameter
        if hasattr(scope, 'callback'):
            callback_strs = [str(arg) for arg in scope.callback]
            if str(first) in callback_strs:
                return self.node

        # Try to retrieve from scope
        scoped = scope.retrieve(self) if hasattr(scope, 'retrieve') else None
        if scoped:
            return scoped.resolve()

        # Handle computed member expressions
        node = copy.deepcopy(self.node)
        if node.get("computed"):
            prop = node.get("property", {})
            property_id = Identifier(copy.deepcopy(prop))
            resolved = property_id.resolve(scope)
            if resolved:
                node["property"] = resolved

        # Check if this references a graph entry
        # This would need to be implemented based on the graph system
        # For now, return the node
        return node

    def graph(self, scope: Any | None = None) -> Optional['Identifier']:
        """
        Get graph representation of this identifier.
        
        Args:
            scope: The scope object to check against
            
        Returns:
            Graph identifier or None
        """
        first = self.first
        if not first:
            return None

        # Check if this is a callback parameter
        if scope and hasattr(scope, 'callback'):
            callback_strs = [str(arg) for arg in scope.callback]
            if str(first) in callback_strs:
                return None

        # This would need integration with the graph system
        # For now, return a copy without builtins
        return self._remove_builtins()

    def _remove_builtins(self) -> 'Identifier':
        """Remove builtin properties like 'length' from identifier chain."""
        current = copy.deepcopy(self.node)

        # Remove .length property chains
        while (current.get("property", {}).get("name") == "length" and
               current.get("type") == "MemberExpression"):
            current = current.get("object", {})

        return Identifier(current)

    def walk(self) -> list[Node]:
        """Walk the identifier and return list of nodes."""
        return [self]

    def __iter__(self) -> Iterator[dict[str, Optional['Identifier']]]:
        """
        Iterator that yields left-right pairs traversing the identifier chain.
        
        Yields:
            Dict with 'left' and 'right' keys containing Identifier objects
        """
        parts = []
        left = self
        right = None

        while left.object:
            last = left.last
            if last:
                if right is None:
                    right = Identifier(last.node)
                else:
                    # Append last.node to right.node
                    right = Identifier(append(last.node, right.node))

                left = left.object
                parts.append({"left": left, "right": right})

        # Return iterator that yields in reverse order
        return iter(reversed(parts))

File: typescript/src/lang/ast/Identifier.ts
import { append, root } from "../estree/estree";
import Node from "./Node";
import _ from "lodash";
import { retrieve } from "../../graph";

interface IdentifierNode {
  type: string;
  [key: string]: any;
}

export default class Identifier extends Node {
  declare node: IdentifierNode;

  constructor(node: IdentifierNode | string) {
    super(node);
  }

  static get types(): string[] {
    return ["Identifier", "MemberExpression", "ThisExpression"];
  }

  override get first(): Identifier | null {
    const n = this.node;
    if (["Identifier", "ThisExpression", "Literal"].includes(n.type)) {
      return new Identifier(n);
    } else if (n.type === "MemberExpression") {
      const r = root(n);
      if (!r.object) return null;
      return new Identifier(r.object);
    }
    return null;
  }

  // Add a getter that returns a string representation of first
  get firstAsString(): string {
    const first = this.first;
    return first ? first.toString() : "";
  }

  // Make toString() explicit for TypeScript
  override toString(): string {
    if (this.node.type === "Identifier") {
      return this.node.name;
    } else if (this.node.type === "ThisExpression") {
      return "this";
    } else if (this.node.type === "MemberExpression") {
      const object = new Identifier(this.node.object);
      const property = new Identifier(this.node.property);
      return `${object.toString()}.${property.toString()}`;
    } else if (this.node.type === "Literal") {
      return String(this.node.value);
    }
    return "";
  }

  override set first(first: Identifier) {
    const f = this.first;
    if (!f) throw new Error("Cannot set first on non-MemberExpression");
    Object.keys(this.node).forEach((key) => delete (this.node as any)[key]);
    Object.assign(this.node, first.node);
  }

  override get object(): Identifier | null {
    if (["Identifier", "ThisExpression", "Literal"].includes(this.node.type)) {
      return null;
    } else if (this.node.type === "MemberExpression") {
      return new Identifier((this.node as any).object);
    }
    return null;
  }

  override set object(object: Identifier) {
    const obj = this.object;
    if (!obj) throw new Error("Cannot set object on non-MemberExpression");
    (this.node as any).object = object.node;
  }

  override get last(): Identifier | null {
    if (["Identifier", "ThisExpression", "Literal"].includes(this.node.type)) {
      return new Identifier(this.node);
    } else if (this.node.type === "MemberExpression") {
      return new Identifier((this.node as any).property);
    }
    return null;
  }

  override set last(last: Identifier) {
    const l = this.last;
    if (!l) throw new Error("Cannot set last on non-MemberExpression");
    if (this.node.type === "MemberExpression") {
      (this.node as any).property = last.node;
    } else {
      Object.keys(this.node).forEach((key) => delete (this.node as any)[key]);
      Object.assign(this.node, last.node);
    }
  }

  override resolve(scope?: any): any {
    if (!scope) return this.node;

    const first = this.first;
    if (!first) return null;

    if (
      scope.callback
        .map((arg: any) => arg.toString())
        .includes(first.toString())
    ) {
      return this.node;
    }

    const scoped = scope.retrieve(this);
    if (scoped) {
      return scoped.resolve();
    }

    let node: any = this.node;
    if ((this.node as any).computed) {
      node = _.cloneDeep(this.node);
      const property = new Identifier(_.cloneDeep((node as any).property));
      const resolved = property.resolve(scope);
      if (resolved) {
        Object.keys((node as any).property).forEach(
          (key) => delete ((node as any).property as any)[key]
        );
        Object.assign((node as any).property, resolved);
      }
    }

    if (retrieve(first)) {
      const state: IdentifierNode = { type: "Identifier", name: "state" };
      return append(state, node);
    }
    return node;
  }

  override graph(scope?: any): Identifier | null {
    const first = this.first;
    if (!first) return null;
    if (
      scope.callback
        .map((arg: any) => arg.toString())
        .includes(first.toString())
    ) {
      return null;
    }
    const rec = retrieve(first);
    if (rec) {
      return removeBuiltins(this);
    }
    return null;
  }

  override walk(): Node[] {
    return [this];
  }

  [Symbol.iterator](): Iterator<{
    left: Identifier;
    right: Identifier | null;
  }> {
    const list: { left: Identifier; right: Identifier | null }[] = [];
    let left: Identifier = this;
    let right: Identifier | null = null;

    while (left.object) {
      const last = left.last!;

      // Handle the case when right is null
      if (right === null) {
        right = new Identifier(last.node);
      } else {
        // When right is not null, append last.node to right.node
        const rightNode = right.node;
        right = new Identifier(append(last.node, rightNode));
      }

      left = left.object!;
      list.push({ left, right });
    }

    let index = list.length - 1;
    return {
      next(): IteratorResult<{ left: Identifier; right: Identifier | null }> {
        if (index >= 0) {
          return { done: false, value: list[index--] };
        }
        return { done: true, value: undefined as any };
      },
    };
  }
}

function removeBuiltins(identifier: Identifier): Identifier {
  let current: any = _.cloneDeep(identifier.node);
  while (current.property?.name === "length") {
    current = current.object;
  }
  return new Identifier(current);
}
